// @serenity-js/web/lib/screenplay/interactions/Enter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enter = void 0;
const io_1 = require("@serenity-js/core/lib/io");
const PageElementInteraction_1 = require("./PageElementInteraction");
/**
 * @desc
 *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
 *  enter a value into a [form `input`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) field.
 *
 * @example <caption>Example widget</caption>
 *  <form>
 *    <input type="text" name="example" id="example" />
 *  </form>
 *
 * @example <caption>Lean Page Object describing the widget</caption>
 *  import { by, Target } from '@serenity-js/webdriverio';
 *
 *  class Form {
 *      static exampleInput = Target.the('example input')
 *          .located(by.id('example'));
 *  }
 *
 * @example <caption>Entering the value into a form field</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Enter } from '@serenity-js/webdriverio';
 *
 *  actorCalled('Esme')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Enter.theValue('Hello world!').into(Form.exampleInput),
 *      );
 *
 * @see {@link Target}
 *
 * @extends {ElementInteraction}
 */
class Enter extends PageElementInteraction_1.PageElementInteraction {
    /**
     * @param {Array<Answerable<string | number | string[] | number[]>>} values
     *  The value to be entered
     *
     * @param {Answerable<PageElement>} field
     *  The field to enter the value into
     */
    constructor(values, field /* todo | Question<AlertPromise> | AlertPromise */) {
        super((0, io_1.formatted) `#actor enters ${values.join(', ')} into ${field}`);
        this.values = values;
        this.field = field;
    }
    /**
     * @desc
     *  Instantiates this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {Array<Answerable<string | number | string[] | number[]>>} values
     *  The value to be entered
     *
     * @returns {EnterBuilder}
     */
    static theValue(...values) {
        return {
            into: (field /* todo Question<AlertPromise> | AlertPromise */) => new Enter(values, field),
        };
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {Promise<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    async performAs(actor) {
        const field = await this.resolve(actor, this.field);
        const valuesToEnter = await (0, io_1.asyncMap)(this.values, value => actor.answer(value));
        return field.enterValue(valuesToEnter.flat());
    }
}
exports.Enter = Enter;
