// @serenity-js/webdriverio/lib/screenplay/models/WebdriverIOModalDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebdriverIOModalDialog = void 0;
const web_1 = require("@serenity-js/web");
class WebdriverIOModalDialog extends web_1.ModalDialog {
    constructor(browser) {
        super();
        this.browser = browser;
    }
    accept() {
        return this.browser.acceptAlert();
    }
    dismiss() {
        return this.browser.dismissAlert();
    }
    text() {
        return this.browser.getAlertText();
    }
    async enterValue(value) {
        const text = [].concat(value).join('');
        await this.browser.sendAlertText(text);
        // Puppeteer auto-accepts the dialog upon sendAlertText
        // other implementations require the dialog to be closed
        if (!this.browser.isDevTools) {
            await this.browser.acceptAlert();
        }
    }
    isPresent() {
        return this.browser.getAlertText()
            .then(() => true, error => {
            if (error && (error.name === 'no such alert'
                || error.message.startsWith('no such alert'))) {
                return false;
            }
            throw error;
        });
    }
}
exports.WebdriverIOModalDialog = WebdriverIOModalDialog;
