// @serenity-js/web/lib/screenplay/questions/Attribute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Attribute = void 0;
const core_1 = require("@serenity-js/core");
const models_1 = require("../models");
const ElementQuestion_1 = require("./ElementQuestion");
/**
 * @desc
 *  Returns the value of the given HTML attribute of a given {@link WebElement},
 *  represented by Answerable<{@link @wdio/types~Element}>
 *
 * @example <caption>Example widget</caption>
 *  <ul id="shopping-list" data-items-left="2">
 *    <li data-state="bought">Coffee<li>
 *    <li data-state="buy">Honey<li>
 *    <li data-state="buy">Chocolate<li>
 *  </ul>
 *
 * @example <caption>Retrieve a HTML attribute of a given WebElement</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { Ensure, equals } from '@serenity-js/assertions';
 *  import { Attribute, by, BrowseTheWeb, Target } from '@serenity-js/webdriverio';
 *
 *  const shoppingList = () =>
 *      Target.the('shopping list').located(by.id('shopping-list'))
 *
 *  actorCalled('Lisa')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Ensure.that(Attribute.called('data-items-left').of(shoppingList()), equals('2')),
 *      )
 *
 * @example <caption>Find WebElements with a given attribute</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { Ensure, includes } from '@serenity-js/assertions';
 *  import { Attribute, BrowseTheWeb, by, Target } from '@serenity-js/webdriverio';
 *
 *  class ShoppingList {
 *      static items = () =>
 *          Target.all('items')
 *              .located(by.css('#shopping-list li'))
 *
 *      static outstandingItems = () =>
 *          ShoppingList.items
 *              .where(Attribute.called('data-state'), includes('buy'))
 *  }
 *
 *  actorCalled('Lisa')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Ensure.that(
 *              Text.ofAll(ShoppingList.outstandingItems()),
 *              equals([ 'Honey', 'Chocolate' ])
 *          ),
 *      )
 *
 * @extends {ElementQuestion}
 * @implements {@serenity-js/core/lib/screenplay/questions~MetaQuestion}
 */
class Attribute extends ElementQuestion_1.ElementQuestion {
    /**
     * @param {Answerable<string>} name
     * @param {@serenity-js/core/lib/screenplay~Answerable<Element>} [element]
     */
    constructor(name, element) {
        super(`"${name}" attribute of ${element}`);
        this.name = name;
        this.element = element;
    }
    /**
     * @param {Answerable<string>} name
     * @returns {Attribute}
     */
    static called(name) {
        return new Attribute(name);
    }
    /**
     * @desc
     *  Resolves to the value of a HTML attribute of the `target` element,
     *  located in the context of a `parent` element.
     *
     * @param {Answerable<PageElement>} parent
     * @returns {Question<Promise<string[]>>}
     *
     * @see {@link Target.all}
     * @see {@link @serenity-js/core/lib/screenplay/questions~MetaQuestion}
     */
    of(parent) {
        return new Attribute(this.name, this.element
            ? models_1.PageElement.of(this.element, parent)
            : parent);
    }
    async answeredBy(actor) {
        if (!this.element) {
            throw new core_1.LogicError(`Target not specified`); // todo: better error message?
        }
        const element = await actor.answer(this.element);
        const name = await actor.answer(this.name);
        return element.attribute(name);
    }
}
exports.Attribute = Attribute;
