// @serenity-js/web/lib/screenplay/questions/Text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Text = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
const models_1 = require("../models");
const ElementQuestion_1 = require("./ElementQuestion");
/**
 * @desc
 *  Resolves to the visible (i.e. not hidden by CSS) `innerText` of:
 *  - a given {@link WebElement}, represented by Answerable<{@link @wdio/types~Element}>
 *  - a group of {@link WebElement}s, represented by Answerable<{@link @wdio/types~ElementList}>
 *
 *  The result includes the visible text of any sub-elements, without any leading or trailing whitespace.
 *
 * @example <caption>Example widget</caption>
 *  <h1>Shopping list</h1>
 *  <ul id="shopping-list">
 *    <li>Coffee<li>
 *    <li class="bought">Honey<li>
 *    <li>Chocolate<li>
 *  </ul>
 *
 * @example <caption>Retrieve text of a single element</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { Ensure, equals } from '@serenity-js/assertions';
 *  import { BrowseTheWeb, by, Target, Text } from '@serenity-js/webdriverio';
 *
 *  const header = () =>
 *      Target.the('header').located(by.tagName('h1'))
 *
 *  actorCalled('Lisa')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Ensure.that(Text.of(header()), equals('Shopping list')),
 *      )
 *
 * @example <caption>Retrieve text of a multiple elements</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { Ensure, equals } from '@serenity-js/assertions';
 *  import { BrowseTheWeb, by, Target, Text } from '@serenity-js/webdriverio';
 *
 *  const shoppingListItems = () =>
 *      Target.the('shopping list items').located(by.css('#shopping-list li'))
 *
 *  actorCalled('Lisa')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Ensure.that(
 *              Text.ofAll(shoppingListItems()),
 *              equals([ 'Coffee', 'Honey', 'Chocolate' ])
 *          ),
 *      )
 *
 * @example <caption>Find element with matching text</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { contain, Ensure } from '@serenity-js/assertions';
 *  import { BrowseTheWeb, by, CssClasses, Target, Text } from '@serenity-js/webdriverio';
 *
 *  const shoppingListItemCalled = (name: string) =>
 *      Target.the('shopping list items').located(by.css('#shopping-list li'))
 *          .where(Text, equals(name))
 *          .first()
 *
 *  actorCalled('Lisa')
 *      .whoCan(BrowseTheWeb.using(browser))
 *      .attemptsTo(
 *          Ensure.that(
 *              CssClasses.of(shoppingListItemCalled('Honey)),
 *              contain('bought')
 *          ),
 *      )
 *
 * @public
 * @see {@link Target}
 */
class Text {
    /**
     * @desc
     *  Retrieves text of a single {@link WebElement},
     *  represented by Answerable<{@link @wdio/types~Element}>.
     *
     * @param {Answerable<PageElement>} element
     * @returns {Question<Promise<string>> & MetaQuestion<Answerable<PageElement>, Promise<string>>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~MetaQuestion}
     */
    static of(element) {
        return TextOfSingleElement.of(element);
    }
    static ofAll(elements) {
        if (elements instanceof models_1.PageElements) {
            return TextOfMultipleElements.of(elements);
        }
        return core_1.Question.about((0, core_1.d) `the text of ${elements}`, async (actor) => {
            const pageElements = await actor.answer(elements);
            return (0, io_1.asyncMap)(pageElements, element => element.text());
        });
    }
}
exports.Text = Text;
class TextOfSingleElement extends ElementQuestion_1.ElementQuestion {
    constructor(element) {
        super(`the text of ${element}`);
        this.element = element;
    }
    static of(element) {
        return core_1.Question.createAdapter(new TextOfSingleElement(element));
    }
    of(parent) {
        return new TextOfSingleElement(models_1.PageElement.of(this.element, parent));
    }
    async answeredBy(actor) {
        const element = await actor.answer(this.element);
        return element.text();
    }
}
class TextOfMultipleElements extends ElementQuestion_1.ElementQuestion {
    constructor(elements) {
        super((0, core_1.d) `the text of ${elements}`);
        this.elements = elements;
    }
    static of(elements) {
        return core_1.Question.createAdapter(new TextOfMultipleElements(elements));
    }
    of(parent) {
        return new TextOfMultipleElements(this.elements.of(parent));
    }
    async answeredBy(actor) {
        const elements = await actor.answer(this.elements);
        return (0, io_1.asyncMap)(elements, element => element.text());
    }
}
