// @serenity-js/webdriverio/lib/adapter/reporter/BrowserCapabilitiesReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserCapabilitiesReporter = void 0;
const events_1 = require("@serenity-js/core/lib/events");
const reporter_1 = require("@wdio/reporter");
const TagPrinter_1 = require("./TagPrinter");
/**
 * @package
 */
class BrowserCapabilitiesReporter extends reporter_1.default {
    constructor(options) {
        super({ ...options, stdout: false });
        this.tagPrinter = new TagPrinter_1.TagPrinter();
        this.tags = [];
        this.serenity = options.serenity;
        this.on('runner:start', BrowserCapabilitiesReporter.prototype.recordBrowserAndPlatformTags.bind(this));
        this.on('test:start', BrowserCapabilitiesReporter.prototype.emitRecordedTags.bind(this));
    }
    recordBrowserAndPlatformTags(event) {
        const tags = event.isMultiremote
            ? this.tagsForAll(event.capabilities) // fixme: WDIO MultiremoteCapabilities seem to have incorrect definition?
            : this.tagPrinter.tagsFor(event.capabilities);
        this.tags.push(...tags);
    }
    tagsForAll(capabilities) {
        return Object.keys(capabilities)
            .reduce((existingTags, entryName) => {
            const newTags = this.tagPrinter.tagsFor(capabilities[entryName]);
            const uniqueNewTags = newTags.filter(tag => !existingTags.some((existingTag) => existingTag.equals(tag)));
            return [
                ...existingTags,
                // todo: maybe add some additional tag to indicate the custom capability name,
                //  or the fact that it's a multi-remote scenario?
                ...uniqueNewTags,
            ];
        }, []);
    }
    emitRecordedTags() {
        this.tags.forEach(tag => {
            this.serenity.announce(new events_1.SceneTagged(this.serenity.currentSceneId(), tag, this.serenity.currentTime()));
        });
    }
}
exports.BrowserCapabilitiesReporter = BrowserCapabilitiesReporter;
