// @serenity-js/web/lib/screenplay/models/PageElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageElement = void 0;
const core_1 = require("@serenity-js/core");
const tiny_types_1 = require("tiny-types");
const abilities_1 = require("../abilities");
class PageElement {
    constructor(locator) {
        this.locator = locator;
        (0, tiny_types_1.ensure)('native element locator', locator, (0, tiny_types_1.isDefined)());
    }
    static located(selector) {
        return core_1.Question.about((0, core_1.d) `page element located ${selector}`, async (actor) => {
            const bySelector = await actor.answer(selector);
            return abilities_1.BrowseTheWeb.as(actor).locate(bySelector).element();
        });
    }
    // todo: review usages and consider removing if not used
    static of(childElement, parentElement) {
        return core_1.Question.about((0, core_1.d) `${childElement} of ${parentElement})`, async (actor) => {
            const child = await actor.answer(childElement);
            const parent = await actor.answer(parentElement);
            return child.of(parent);
        });
    }
    async nativeElement() {
        return this.locator.nativeElement();
    }
    toString() {
        return `PageElement located ${this.locator.toString()}`;
    }
}
exports.PageElement = PageElement;
